### PR TITLE
Add protobuf definitions for API resources

### LINKED ISSUE #1605: Protobuf generation is missing
**What happened**:
Using Gardener on a 1.16 Kubernetes cluster does not work properly because the kube-controller-manager only uses `protobuf` in 1.16, and the Gardener does not yet support this content type.

**What you expected to happen**:
Gardener should provide protobuf definitons.

**How to reproduce it (as minimally and precisely as possible)**:
Deploy Gardener on 1.16 cluster and try to delete a namespace. See error logs in the kube-controller-manager like:

```
E1111 09:24:51.733918       1 namespace_controller.go:148] deletion of namespace extension-provider-openstack-7ld84 failed: [object *v1beta1.ShootList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1alpha1.ShootList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1beta1.SecretBindingList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1alpha1.OpenIDConnectPresetList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1beta1.QuotaList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1alpha1.SecretBindingList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1alpha1.PlantList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1alpha1.QuotaList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message, object *v1alpha1.BackupEntryList does not implement the protobuf marshalling interface and cannot be encoded to a protobuf message]
```

**Environment**:

- Gardener version: 0.31.3
- Kubernetes version (use `kubectl version`): 1.16.2
- Cloud provider or hardware configuration: AWS

### PR DESCRIPTION
**Pull request by ialidzhikov**
**What this PR does / why we need it**:
Add protobuf definitions for API resources.

**Which issue(s) this PR fixes**:
Fixes #1605

**Special notes for your reviewer**:
This PR requires:
- [x] merge of gardener/gardener#2104 

**Release note**:

```noteworthy user
Gardener does now have protobuf definitions for its API resources. Clients can now specify `application/vnd.kubernetes.protobuf` content type in the Accept header for their requests. For more information about protobuf representation see this [documentation](https://kubernetes.io/docs/reference/using-api/api-concepts/#alternate-representations-of-resources).
```

```noteworthy operation
An issue preventing namespace_controller to delete namespaces in the Garden cluster for Kubernetes versions >= v1.16 is now fixed.
```

### CODE REVIEW

File: hack/api-reference/core.md [hunk 1]
```diff
@@ -15,22 +15,6 @@ Resource Types:
 <a href="#core.gardener.cloud/v1beta1.BackupEntry">BackupEntry</a>
 </li><li>
 <a href="#core.gardener.cloud/v1beta1.CloudProfile">CloudProfile</a>
-</li><li>
```
Comment (rfranzke) [inline hack/api-reference/core.md]: Why do we have these changes?

File: pkg/apis/core/v1alpha1/types_common.go [hunk 2]
```diff
@@ -33,16 +33,16 @@ const (
...
-	Description string `json:"description"`
+	Description string `json:"description" protobuf:"bytes,1,opt,name=description"`
 	// ID of the task which caused this last error
 	// +optional
-	TaskID *string `json:"taskID,omitempty"`
+	TaskID *string `json:"taskID,omitempty" protobuf:"bytes,2,opt,name=taskID"`
 	// Well-defined error codes of the last error(s).
 	// +optional
-	Codes []ErrorCode `json:"codes,omitempty"`
+	Codes []ErrorCode `json:"codes,omitempty" protobuf:"bytes,3,rep,name=codes,casttype=ErrorCode"`
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_common.go:42]: When is `casttype` needed? Sometimes you use it, sometimes not.

File: pkg/apis/core/v1alpha1/types_utils.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ const (
 // https://github.com/kubernetes/kubernetes/issues/55890
 // https://github.com/kubernetes-sigs/cluster-api/issues/137
 type ProviderConfig struct {
-	runtime.RawExtension `json:",inline"`
+	runtime.RawExtension `json:",inline" protobuf:"bytes,1,opt,name=rawExtension"`
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_utils.go:33]: Is this correct (especially `name=rawExtension`)? Why is not it not done for all inline types (e.g., `TypeMeta`)?

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 4]
```diff
@@ -147,24 +147,24 @@ type ShootStatus struct {
 type Addons struct {
 	// KubernetesDashboard holds configuration settings for the kubernetes dashboard addon.
 	// +optional
-	KubernetesDashboard *KubernetesDashboard `json:"kubernetes-dashboard,omitempty"`
+	KubernetesDashboard *KubernetesDashboard `json:"kubernetes-dashboard,omitempty" protobuf:"bytes,1,opt,name=kubernetesDashboard"`
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_shoot.go:150]: Are the names here correct (`kubernetes-dashboard` vs. `kubernetesDashboard`)?

File: pkg/apis/core/v1alpha1/types_shoot.go [hunk 5]
```diff
@@ -383,120 +383,120 @@ type ServiceAccountConfig struct {
...
 	// Defaults to URI of the API server.
 	// +optional
-	Issuer *string `json:"issuer,omitempty"`
+	Issuer *string `json:"issuer,omitempty" protobuf:"bytes,1,opt,name=issuer"`
 	// SigningKeySecret is a reference to a secret that contains an optional private key of the
 	// service account token issuer. The issuer will sign issued ID tokens with this private key.
 	// Only useful if service account tokens are also issued by another external system.
 	// +optional
-	SigningKeySecret *corev1.LocalObjectReference `json:"signingKeySecretName,omitempty"`
+	SigningKeySecret *corev1.LocalObjectReference `json:"signingKeySecretName,omitempty" protobuf:"bytes,2,opt,name=signingKeySecretName"`
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_shoot.go:391]: Again, sometimes we have `castType`, sometimes we don't. What's the policy?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/apis/core/v1alpha1/types_shoot.go (hunk 4 above)
Comment (ialidzhikov) [reply pkg/apis/core/v1alpha1/types_shoot.go:150]: `-` is not  allowed in proto message field name. Ref https://developers.google.com/protocol-buffers/docs/reference/proto2-spec

File: pkg/apis/core/v1alpha1/types_shoot.go (hunk 5 above)
Comment (ialidzhikov) [reply pkg/apis/core/v1alpha1/types_shoot.go:391]: I am not adding the protobuf tags manually, they are added automatically by the generator. But I will try to check what is the rationale behind.

Comment (rfranzke) [reply pkg/apis/core/v1alpha1/types_shoot.go:391]: Ah okay, even better. I was afraid that we have to maintain this manually... 😄

Comment (ialidzhikov) [reply pkg/apis/core/v1alpha1/types_shoot.go:391]: `casttype` is used to change the generated field type. It allows to have proto message such as

```proto
// KubeProxyConfig contains configuration settings for the kube-proxy.
message KubeProxyConfig {
  optional KubernetesConfig kubernetesConfig = 1;

  // Mode specifies which proxy mode to use.
  // defaults to IPTables.
  // +optional
  optional string mode = 2;
}
```
where `mode` type is `string`.
The `mode` in the go type is of type `ProxyMode`

```golang
// KubeProxyConfig contains configuration settings for the kube-proxy.
type KubeProxyConfig struct {
	KubernetesConfig `json:",inline" protobuf:"bytes,1,opt,name=kubernetesConfig"`
	// Mode specifies which proxy mode to use.
	// defaults to IPTables.
	// +optional
	Mode *ProxyMode `json:"mode,omitempty" protobuf:"bytes,2,opt,name=mode,casttype=ProxyMode"`
}
```

and the generated unmarshal func in `generated.pb.go` properly casts to `ProxyMode`.

You can note that casttype is used for type aliases  - `ProxyMode`, `corev1.ServiceExternalTrafficPolicyType`, etc.

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Review by timebertt — DISMISSED** [review]
Thanks for the PR! /lgtm

I tested it my local setup, bumped the local garden to `v1.17.3` and namespace deletion worked completely fine.
I guess, after we merge this PR, we can bump the nodeless garden version again, right?

**Comment by timuthy** [comment]
> I guess, after we merge this PR, we can bump the nodeless garden version again, right?

Just a revert of https://github.com/gardener/gardener/pull/2075/commits/1ca6ce67e2da83e2a087b9cf86ff2eee2d8401f5

**Comment by rfranzke** [comment]
@ialidzhikov can you rebase please so that we can get this one merged? :)

**Comment by rfranzke** [comment]
> https://github.com/gardener/gardener/pull/2113#issuecomment-606659096

@timuthy you wanna open a revert PR for the mentioned commit? :)

**Comment by timebertt** [comment]
> @timuthy you wanna open a revert PR for the mentioned commit? :)

There you go: https://github.com/gardener/gardener/pull/2125
Already had the PR prepared locally, as I tested @ialidzhikov's change with a nodeless cluster in version `v1.17.3`.

**Review by rfranzke — APPROVED** [review]

**Pull request merged**